File: my-code/740/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/05/28 18: 13
# @Author   : Ranshi
# @File     : 740. 删除并获得点数.py
from typing import List


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        val = max(nums)
        total = [0] * (val + 1)
        for item in nums:
            total[item] += item

        size = len(total)
        first, second = total[0], max(total[0], total[1])
        for i in range(2, size):
            first, second = second, max(first + total[i], second)
        return second


if __name__ == '__main__':
    s = Solution()
    print(s.deleteAndEarn(nums=[2, 2, 3, 3, 3, 4]))

File: my-code/740/main.go
package main

import (
	"fmt"
	"sort"
)

// @Time     : 2021/08/08 15:26
// @Author   : Ranshi
// @File     : main.go
func maxInt(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func deleteAndEarn(nums []int) int {
	sort.Ints(nums)
	arr := []int{nums[0]}
	for i := 1; i < len(nums); i++ {
		if nums[i] == nums[i-1] {
			arr[len(arr)-1] += nums[i]
		} else {
			if nums[i] > nums[i-1]+1 {
				arr = append(arr, 0)
			}
			arr = append(arr, nums[i])
		}
	}
	if len(arr) == 1 {
		return arr[0]
	}
	dp := make([]int, len(arr))
	dp[0] = arr[0]
	dp[1] = maxInt(arr[0], arr[1])
	for i := 2; i < len(arr); i++ {
		dp[i] = maxInt(dp[i-1], arr[i]+dp[i-2])
	}
	return dp[len(dp)-1]
}

func main() {
	fmt.Println(deleteAndEarn([]int{2, 3}))
}
